Framework/Indexing/inc/MantidIndexing/SpectrumNumberTranslator.h

```diff
@@ -0,0 +1,91 @@
+#ifndef MANTID_INDEXING_SPECTRUMNUMBERTRANSLATOR_H_
+#define MANTID_INDEXING_SPECTRUMNUMBERTRANSLATOR_H_
+
+#include "MantidIndexing/DllConfig.h"
+#include "MantidIndexing/IPartitioning.h"
+#include "MantidIndexing/SpectrumIndexSet.h"
+#include "MantidIndexing/SpectrumNumber.h"
+
+#include <unordered_map>
+
+namespace Mantid {
+namespace Indexing {
+
+/** SpectrumNumberTranslator : TODO: DESCRIPTION
+
+  @author Simon Heybrock
+  @date 2016
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_INDEXING_DLL SpectrumNumberTranslator {
+public:
+  SpectrumNumberTranslator(const std::vector<SpectrumNumber> &spectrumNumbers,
+                           const IPartitioning &partitioning)
+      : m_partition(partitioning.partitionIndex()) {
+    size_t currentIndex = 0;
+    for (size_t i = 0; i < spectrumNumbers.size(); ++i) {
+      auto number = spectrumNumbers[i];
+      auto partition = partitioning.partitionIndexOf(number);
+      m_partitions[number] = partition;
+      if (partition == m_partition)
+        m_indices[number] = currentIndex++;
+    }
+  }
+
+  // Full set
+  SpectrumIndexSet makeIndexSet() { return SpectrumIndexSet(m_indices.size()); }
```

Comment:
@OwenArnold Yes, if only the size is set we basically have a low-overhead `IndexSet` that represents "all indices". In that case `IndexSet` is just a 1:1 mapping from input to output, without the need to access any index vectors.